[clang][DebugInfo] Emit global variable definitions for static data members with constant initializers

Michael137 opened the pull request:
When an LLDB user asks for the value of a static data member, LLDB starts by
searching the Names accelerator table for the corresponding variable definition
DIE. For static data members with out-of-class definitions that works fine,
because those get represented as global variables with a location and making them
eligible to be added to the Names table. However, in-class definitions won’t get
indexed because we usually don't emit global variables for them. So in DWARF
we end up with a single `DW_TAG_member` that usually holds the constant initializer.
But we don't get a corresponding CU-level `DW_TAG_variable` like we do for
out-of-class definitions.

To make it more convenient for debuggers to get to the value of inline static data members,
this patch makes sure we emit definitions for static variables with constant initializers
the same way we do for other static variables. This also aligns Clang closer to GCC, which
produces CU-level definitions for inline statics and also emits these into `.debug_pubnames`.

The implementation keeps track of newly created static data members. Then in
`CGDebugInfo::finalize`, we emit a global `DW_TAG_variable` with a `DW_AT_const_value` for
any of those declarations that didn't end up with a definition in the `DeclCache`.

The newly emitted `DW_TAG_variable` will look as follows:
```
0x0000007b:   DW_TAG_structure_type
                DW_AT_calling_convention        (DW_CC_pass_by_value)
                DW_AT_name      ("Foo")
                ...

0x0000008d:     DW_TAG_member
                  DW_AT_name    ("i")
                  DW_AT_type    (0x00000062 "const int")
                  DW_AT_external        (true)
                  DW_AT_declaration     (true)
                  DW_AT_const_value     (4)

Newly added
vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

0x0000009a:   DW_TAG_variable
                DW_AT_specification     (0x0000008d "i")
                DW_AT_const_value       (4)
                DW_AT_linkage_name      ("_ZN2t2IiE1iIfEE")
```

Michael137 committed:
[clang][DebugInfo] Emit global variable definitions for static data members with constant initializers

When an LLDB user asks for the value of a static data member, LLDB starts by
searching the Names accelerator table for the corresponding variable definition
DIE. For static data members with out-of-class definitions that works fine,
because those get represented as global variables with a location and making them
eligible to be added to the Names table. However, in-class definitions won’t get
indexed because we usually don't emit global variables for them. So in DWARF
we end up with a single `DW_TAG_member` that usually holds the constant initializer.
But we don't get a corresponding CU-level `DW_TAG_variable` like we do for
out-of-class definitions.

To make it more convenient for debuggers to get to the value of inline static data members,
this patch makes sure we emit definitions for static variables with constant initializers
the same way we do for other static variables. This also aligns Clang closer to GCC, which
produces CU-level definitions for inline statics and also emits these into `.debug_pubnames`.

The implementation keeps track of newly created static data members. Then in
`CGDebugInfo::finalize`, we emit a global `DW_TAG_variable` with a `DW_AT_const_value` for
any of those declarations that didn't end up with a definition in the `DeclCache`.

The newly emitted `DW_TAG_variable` will look as follows:
```
0x0000007b:   DW_TAG_structure_type
                DW_AT_calling_convention        (DW_CC_pass_by_value)
                DW_AT_name      ("Foo")
                ...

0x0000008d:     DW_TAG_member
                  DW_AT_name    ("i")
                  DW_AT_type    (0x00000062 "const int")
                  DW_AT_external        (true)
                  DW_AT_declaration     (true)
                  DW_AT_const_value     (4)

Newly added
vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

0x0000009a:   DW_TAG_variable
                DW_AT_specification     (0x0000008d "i")
                DW_AT_const_value       (4)
                DW_AT_linkage_name      ("_ZN2t2IiE1iIfEE")
```

Michael137 commented:
Summing the size of all *.o files in my local debug Clang build increased by `0.7%` with this patch

Michael137 committed:
fixup! add test-case for explicitly inline member

Michael137 committed:
fixup! clang-format

Michael137 commented:
This could also help with the type merging issues discussed in https://github.com/llvm/llvm-project/pull/68721#issuecomment-1764685418

avl-llvm commented:
Should not we remove constant initializer from the type definition if we have a DW_TAG_variable with a DW_AT_const_value now?

```
0x0000008d:     DW_TAG_member
                  DW_AT_name    ("i")
                  DW_AT_type    (0x00000062 "const int")
                  DW_AT_external        (true)
                  DW_AT_declaration     (true)
                  DW_AT_const_value     (4)   <<<<<<<<<<<<<<<<<<<<<
```

dwblaikie commented on clang/test/CodeGenCXX/debug-info-static-inline-member.cpp:
```diff
@@ -0,0 +1,87 @@
+// CHECK:      @{{.*}}cexpr_empty{{.*}} = 
+// CHECK-SAME    !dbg ![[EMPTY_GLOBAL:[0-9]+]]
+
+// CHECK:      !DIGlobalVariableExpression(var: ![[INT_VAR:[0-9]+]], expr: !DIExpression())
```
Why doesn't this have a value/expression? (but INT_VAR2 does)

Michael137 replied:
Because this one has a corresponding `llvm::GlobalVariable` which doesn't go down the code-path I added. We emit a `DW_AT_location` only for this. Should we be emitting both a location and constant in this case?

dwblaikie replied:
Ah, that was the intent (the line 27 taking address-of)? Maybe give the variable a more descriptive name to make it clear this is testing the case where the variable has a definition emitted to the IR

dwblaikie commented:
> Should not we remove constant initializer from the type definition if we have a DW_TAG_variable with a DW_AT_const_value now?

Yep, +1 to this. This is where the type normalization benefit would come from - no longer having this const value on the member declaration, but only on the definition.

dwblaikie commented:
> Summing the size of all *.o files in my local debug Clang build increased by 0.7% with this patch

That's a bit significant. Any chance you could get a per-section breakdown/growth on that? (& exact flags - is this with compressed debug info, for instance? Or not?) I use `bloaty` for this ( https://github.com/google/bloaty ) - though it's a bit involved to do this to sum over all the object files - summing up the results from per-object file, etc...

Michael137 commented:
> > Should not we remove constant initializer from the type definition if we have a DW_TAG_variable with a DW_AT_const_value now?
> 
> Yep, +1 to this. This is where the type normalization benefit would come from - no longer having this const value on the member declaration, but only on the definition.

Makes sense. I'll create a separate PR for that, unless people are fine with doing that as part of this change

dwblaikie commented:
> > > Should not we remove constant initializer from the type definition if we have a DW_TAG_variable with a DW_AT_const_value now?
> > 
> > 
> > Yep, +1 to this. This is where the type normalization benefit would come from - no longer having this const value on the member declaration, but only on the definition.
> 
> Makes sense. I'll create a separate PR for that, unless people are fine with doing that as part of this change

If it's not too much code might make sense to do it as part of this change - so the size impacts are more clear, for instance.

Michael137 commented:
>  I use `bloaty` for this ( https://github.com/google/bloaty ) - though it's a bit involved to do this to sum over all the object files - summing up the results from per-object file, etc...

```
bloaty `find ./patched-build/lib -name *.o`       │ bloaty `find ./no-patch-build/lib -name *.o `
  FILE SIZE        VM SIZE                        │    FILE SIZE        VM SIZE    
-------------  --------------                     │ --------------  -------------- 
54.6%  2.02Gi  60.3%  2.02Gi    ,__debug_str      │  54.5%  2.01Gi  60.1%  2.01Gi    ,__debug_str
22.0%   835Mi  24.3%   835Mi    ,__debug_info     │  22.1%   831Mi  24.3%   831Mi    ,__debug_info
 6.2%   236Mi   0.0%       0    String Table      │   6.1%   228Mi   0.0%       0    String Table
 4.5%   170Mi   5.0%   170Mi    ,__text           │   4.5%   169Mi   5.0%   169Mi    ,__text
 2.6%  97.6Mi   2.8%  97.6Mi    ,__debug_line     │   2.8%   104Mi   3.1%   104Mi    ,__debug_line
 2.3%  88.0Mi   2.6%  88.0Mi    ,__apple_types    │   2.3%  87.8Mi   2.6%  87.8Mi    ,__apple_types
 2.2%  83.1Mi   2.4%  83.1Mi    ,__apple_names    │   2.1%  81.0Mi   0.0%       0    [Unmapped]
 2.2%  81.8Mi   0.0%       0    [Unmapped]        │   2.1%  78.5Mi   2.3%  78.5Mi    ,__apple_names
 1.7%  65.6Mi   1.9%  65.6Mi    ,__compact_unwind │   1.7%  65.4Mi   1.9%  65.4Mi    ,__compact_unwind
 1.0%  39.2Mi   0.0%       0    Symbol Table      │   1.0%  39.0Mi   0.0%       0    Symbol Table
 0.2%  8.05Mi   0.2%  8.05Mi    ,__const          │   0.2%  7.55Mi   0.2%  7.55Mi    ,__const
 0.2%  7.91Mi   0.2%  7.91Mi    ,__cstring        │   0.2%  7.54Mi   0.2%  7.54Mi    ,__cstring
 0.1%  5.24Mi   0.2%  5.24Mi    ,__debug_abbrev   │   0.1%  5.27Mi   0.2%  5.27Mi    ,__debug_abbrev
 0.1%  2.31Mi   0.0%       0    [Mach-O Headers]  │   0.1%  2.30Mi   0.0%       0    [Mach-O Headers]
 0.0%  1.38Mi   0.0%  1.38Mi    ,__debug_ranges   │   0.0%  1.38Mi   0.0%  1.38Mi    ,__debug_ranges
 0.0%   994Ki   0.0%   994Ki    ,__apple_namespac │   0.0%   949Ki   0.0%   949Ki    ,__apple_namespac
 0.0%       0   0.0%   354Ki    ,__bss            │   0.0%       0   0.0%   354Ki    ,__bss
 0.0%   324Ki   0.0%   324Ki    ,__data           │   0.0%   324Ki   0.0%   324Ki    ,__data
 0.0%   283Ki   0.0%   283Ki    ,__StaticInit     │   0.0%   283Ki   0.0%   283Ki    ,__StaticInit
 0.0%  31.3Ki   0.0%  61.4Ki    [10 Others]       │   0.0%  31.9Ki   0.0%  62.0Ki    [11 Others]
 0.0%  58.6Ki   0.0%  58.6Ki    ,__apple_objc     │   0.0%  58.6Ki   0.0%  58.6Ki    ,__apple_objc
00.0%  3.71Gi 100.0%  3.35Gi    TOTAL             │ 100.0%  3.68Gi 100.0%  3.33Gi    TOTAL
```

Probably the linkage names that contribute most?

> That's a bit significant. Any chance you could get a per-section breakdown/growth on that? (& exact flags - is this with compressed debug info, for instance? Or not?)

This is an example command line invocation (with some paths redacted) for one of the object files:
```
./patched-build/bin/clang++ -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_LIBCPP_ENABLE_HARDENED_MODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__
STDC_LIMIT_MACROS -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissin
g-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstr
ing-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -g -std=c++17 -arch arm64 -fno-exceptions -funwind-tables -fno-rtti"    
```

Michael137 committed:
[clang][DebugInfo] Stop emitting DW_AT_const_value on static member declarations

Since `f07cf5aaf96a70123530239d07c1e655dc91b98d` we emit a definition
for any static data member with a constant initializer. That definition
will either have a location or constant attached to it. So having the
constant on the declaration is redundant. It also helps with the
type merging inconsistencies encountered in
https://github.com/llvm/llvm-project/pull/68721

Michael137 commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -1677,22 +1677,13 @@ CGDebugInfo::CreateRecordStaticField(const VarDecl *Var, llvm::DIType *RecordTy,
   auto Align = getDeclAlignIfRequired(Var, CGM.getContext());
   llvm::DIDerivedType *GV = DBuilder.createStaticMemberType(
-      RecordTy, VName, VUnit, LineNumber, VTy, Flags, C, Align);
+      RecordTy, VName, VUnit, LineNumber, VTy, Flags, /* Val */ nullptr, Align);
```
This parameter on `createStaticMemberType` is technically not needed anymore. So could think about removing it in a follow-up

dwblaikie commented:
size report sounds generally OK - but there's a bunch of what look like unrelated or strange results in there, perhaps they weren't compared at exactly the same version? (like why did the apple_names size go down? How did the .text and .debuG_line size change at all? )

dwblaikie commented on clang/test/CodeGenCXX/debug-info-static-member.cpp:
```diff
@@ -67,15 +67,15 @@ int C::a = 4;
 // CHECK-NOT:            offset:
-// CHECK-SAME:           flags: DIFlagStaticMember,
-// CHECK-SAME:           extraData: i1 true)
+// CHECK-SAME:           flags: DIFlagStaticMember
+// CHECK-NOT:            extraData:
```
Perhaps this test case could be extended to check the constatn values on the definition metadata? & /maybe/ we don't need a new test case? (or maybe we do, but perhaps they could be added to this file, rather than another one?)

Michael137 replied:
Merging the two makes sense. I tried that initially but the C++98 FileCheck was tedious to work around. But with the latest iteration of the patch that shouldn't be a problem

dwblaikie commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -5883,6 +5907,18 @@ void CGDebugInfo::finalize() {
+  for (auto const *VD : StaticDataMemberDefinitionsToEmit) {
+    assert(VD->isStaticDataMember());
+
+    if (auto It = DeclCache.find(VD); It != DeclCache.end())
```
DeclCache.contains might be simpler to use/read here?

clayborg reviewed (commented):
One question I have here: where will the DW_TAG_variable get emitted for these `constexpr`? For actual static member variables we emit a single DW_TAG_variable in the file that declares the global variable, but for `constexpr` we won't be able to do this will we? Will we end up emitting a new copy each time someone include the header file? In that case we might end up with many global variables being defined in a linked executable that includes this header file more than once?

dwblaikie reviewed (approved):
Few minor issues, but looks good to me.
(you metnioned somewhere an lldb issue I think with this patch when the value is removed from the static member declaration inside the class? If that's a problem - probably hold off on committing this until lldb's been fixed so this doesn't regress things? And document the dependence clearly in the commit message)

Michael137 commented:
> size report sounds generally OK - but there's a bunch of what look like unrelated or strange results in there, perhaps they weren't compared at exactly the same version? (like why did the apple_names size go down? How did the .text and .debuG_line size change at all? )

Yea I was confused about that too. Might've mixed up commits of the control build. Did measurements again and here's what I got:
```
$ bloaty `find ./patched-build/lib -name *.o `      │ $ bloaty `find ./no-patch-build/lib -name *.o `
    FILE SIZE        VM SIZE                        │    FILE SIZE        VM SIZE    
 --------------  --------------                     │ --------------  -------------- 
  54.6%  2.02Gi  60.3%  2.02Gi    ,__debug_str      │  54.5%  2.01Gi  60.3%  2.01Gi    ,__debug_str
  22.0%   834Mi  24.3%   834Mi    ,__debug_info     │  22.1%   833Mi  24.4%   833Mi    ,__debug_info
   6.2%   236Mi   0.0%       0    String Table      │   6.2%   236Mi   0.0%       0    String Table
   4.5%   170Mi   5.0%   170Mi    ,__text           │   4.5%   170Mi   5.0%   170Mi    ,__text
   2.6%  97.6Mi   2.8%  97.6Mi    ,__debug_line     │   2.6%  98.6Mi   2.9%  98.6Mi    ,__debug_line
   2.3%  88.0Mi   2.6%  88.0Mi    ,__apple_types    │   2.3%  88.0Mi   2.6%  88.0Mi    ,__apple_types
   2.2%  83.1Mi   2.4%  83.1Mi    ,__apple_names    │   2.2%  81.8Mi   0.0%       0    [Unmapped]
   2.2%  81.8Mi   0.0%       0    [Unmapped]        │   2.1%  77.8Mi   2.3%  77.8Mi    ,__apple_names
   1.7%  65.6Mi   1.9%  65.6Mi    ,__compact_unwind │   1.7%  65.6Mi   1.9%  65.6Mi    ,__compact_unwind
   1.0%  39.2Mi   0.0%       0    Symbol Table      │   1.0%  39.2Mi   0.0%       0    Symbol Table
   0.2%  8.05Mi   0.2%  8.05Mi    ,__const          │   0.2%  8.05Mi   0.2%  8.05Mi    ,__const
   0.2%  7.91Mi   0.2%  7.91Mi    ,__cstring        │   0.2%  7.92Mi   0.2%  7.92Mi    ,__cstring
   0.1%  5.05Mi   0.1%  5.05Mi    ,__debug_abbrev   │   0.1%  5.22Mi   0.2%  5.22Mi    ,__debug_abbrev
   0.1%  2.31Mi   0.0%       0    [Mach-O Headers]  │   0.1%  2.31Mi   0.0%       0    [Mach-O Headers]
   0.0%  1.38Mi   0.0%  1.38Mi    ,__debug_ranges   │   0.0%  1.38Mi   0.0%  1.38Mi    ,__debug_ranges
   0.0%   994Ki   0.0%   994Ki    ,__apple_namespac │   0.0%   994Ki   0.0%   994Ki    ,__apple_namespac
   0.0%       0   0.0%   354Ki    ,__bss            │   0.0%       0   0.0%   354Ki    ,__bss
   0.0%   324Ki   0.0%   324Ki    ,__data           │   0.0%   324Ki   0.0%   324Ki    ,__data
   0.0%   283Ki   0.0%   283Ki    ,__StaticInit     │   0.0%   283Ki   0.0%   283Ki    ,__StaticInit
   0.0%  31.3Ki   0.0%  61.4Ki    [10 Others]       │   0.0%  31.3Ki   0.0%  61.4Ki    [10 Others]
   0.0%  58.6Ki   0.0%  58.6Ki    ,__apple_objc     │   0.0%  58.6Ki   0.0%  58.6Ki    ,__apple_objc
 100.0%  3.70Gi 100.0%  3.35Gi    TOTAL             │ 100.0%  3.69Gi 100.0%  3.34Gi    TOTAL
```

Section breakdown is still similar

clayborg commented:
> Few minor issues, but looks good to me. (you metnioned somewhere an lldb issue I think with this patch when the value is removed from the static member declaration inside the class? If that's a problem - probably hold off on committing this until lldb's been fixed so this doesn't regress things? And document the dependence clearly in the commit message)

I saw the patch to lldb to test that we can find global variables that are `constexpr`. In upstream, we can't use find global variables, but the expression parser does work for these. We should make sure that both `FindGlobalVariables(...)` and expression parsing work for all `constexpr` globals.

Michael137 commented:
> > Few minor issues, but looks good to me. (you metnioned somewhere an lldb issue I think with this patch when the value is removed from the static member declaration inside the class? If that's a problem - probably hold off on committing this until lldb's been fixed so this doesn't regress things? And document the dependence clearly in the commit message)
> 
> I saw the patch to lldb to test that we can find global variables that are `constexpr`. In upstream, we can't use find global variables, but the expression parser does work for these. We should make sure that both `FindGlobalVariables(...)` and expression parsing work for all `constexpr` globals.

Yup, the expression evaluator finds them by looking at the `DW_AT_const_value` on the declaration. But since we're moving that constant onto the definition with this patch, that will break that codepath.

Looking into how best to address that now

Michael137 committed:
fixup! simplify DeclCache check

Michael137 committed:
fixup! tests: add checks for existence of definition

Michael137 committed:
fixup! test: make variable name clearer

clayborg commented:
> > > Few minor issues, but looks good to me. (you metnioned somewhere an lldb issue I think with this patch when the value is removed from the static member declaration inside the class? If that's a problem - probably hold off on committing this until lldb's been fixed so this doesn't regress things? And document the dependence clearly in the commit message)
> > 
> > 
> > I saw the patch to lldb to test that we can find global variables that are `constexpr`. In upstream, we can't use find global variables, but the expression parser does work for these. We should make sure that both `FindGlobalVariables(...)` and expression parsing work for all `constexpr` globals.
> 
> Yup, the expression evaluator finds them by looking at the `DW_AT_const_value` on the declaration. But since we're moving that constant onto the definition with this patch, that will break that codepath.
> 
> Looking into how best to address that now

The DWARFASTParserClang.cpp will try to create the class from the DWARF for the class definition. You will need to find the `DW_TAG_variable` when we are creating the static field if there is no `DW_AT_const_value` in the `DW_TAG_member`. But we also need to support the `DW_AT_const_value` being in the `DW_TAG_member` since older DWARF will be emitted like this. Are we going to emit a `DW_AT_const_expr` now in the `DW_TAG_member`? If so, then we will know that we need to look for the DW_TAG_variable. I don't think clang emitted the `DW_AT_const_expr` attribute before.

The other option it to just see if the expression parser works since global variables are found correctly. Maybe the same will happen in this case?

Michael137 commented:
> The DWARFASTParserClang.cpp will try to create the class from the DWARF for the class definition. You will need to find the DW_TAG_variable when we are creating the static field if there is no DW_AT_const_value in the DW_TAG_member. But we also need to support the DW_AT_const_value being in the DW_TAG_member since older DWARF will be emitted like this.

That's 100% correct. I was thinking, before [this block](https://github.com/llvm/llvm-project/blob/8b91de5d6a3f98dcc00bbd286e339e512f7e3682/lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp#L2909-L2919) where we check for the existence of a `const_value_form`. With this patch such constant won't exist on the declaration. So I was thinking we would try to look for the definition. What's interesting is that with this patch, the expression evaluator successfully finds the DW_TAG_variable which have a location attribute but not if they have a constant instead of a location. It's probably some logic that assumes statics always have a location

> Are we going to emit a DW_AT_const_expr now in the DW_TAG_member? If so, then we will know that we need to look for the DW_TAG_variable. I don't think clang emitted the DW_AT_const_expr attribute before.

That wasn't part of this patch. But would make sense to add (i've noticed GCC adds that attribute)

Michael137 commented:
Hmmm it seems like for `constexpr static` data members we don't get a symbol in the MachO the same way we do for out-of-line `const static`s that have a location:
```
struct Foo {                       
    static const int val1;         
    static constexpr int val2 = 5; 
};                                 
                                   
const int Foo::val1 = 10;          

$ nm a.out                            
0000000100003fa4 S __ZN3Foo4val1E     
0000000100000000 T __mh_execute_header
0000000100003f90 T _main            
```

LLDB can fish out the value out of `val1` by doing a `FindSymbol` on the mangled name. But that doesn't work with `val2` because it's not in the symbol table. @clayborg or @jasonmolenda probably know more about this.

Seems to me like if we want to drop the constant off of the declaration entirely, we need to lookup the variable definition in the accelerator table in DWARFASTParserClang when we're constructing the VarDecls.

Michael137 commented:
> One question I have here: where will the DW_TAG_variable get emitted for these `constexpr`? For actual static member variables we emit a single DW_TAG_variable in the file that declares the global variable, but for `constexpr` we won't be able to do this will we? Will we end up emitting a new copy each time someone include the header file? In that case we might end up with many global variables being defined in a linked executable that includes this header file more than once?

That's true, if defined in a header, we'll emit a `DW_TAG_variable` for the constant in each compile unit the header is included in. GCC does do the right thing and only emit the definition DIE in a single CU. We should probably do the same. Though not sure at which level. Possibly the DWARF linker?

Michael137 committed:
fixup! fix lldb-dap test

dwblaikie commented:
> That's true, if defined in a header, we'll emit a DW_TAG_variable for the constant in each compile unit the header is included in. GCC does do the right thing and only emit the definition DIE in a single CU. We should probably do the same. Though not sure at which level we want to catch that.

Which variable are you discussing here, `val1` or `val2`?

For `val1`, we could not emit the constant value and only emit the real definition (there's /some/ risk here - non-ODR uses (or otherwise optimized away uses) of the variable may mean that the object file that defines the variable won't be linked in - so we'd miss the constant value)
For `val2` the variable is effectively `inline` and doesn't have a home, so there's no one place that we can emit the `DW_TAG_variable` out-of-line definition...

Michael137 commented:
> > That's true, if defined in a header, we'll emit a DW_TAG_variable for the constant in each compile unit the header is included in. GCC does do the right thing and only emit the definition DIE in a single CU. We should probably do the same. Though not sure at which level we want to catch that.
> 
> Which variable are you discussing here, `val1` or `val2`?
> 
> For `val1`, we could not emit the constant value and only emit the real definition (there's /some/ risk here - non-ODR uses (or otherwise optimized away uses) of the variable may mean that the object file that defines the variable won't be linked in - so we'd miss the constant value) For `val2` the variable is effectively `inline` and doesn't have a home, so there's no one place that we can emit the `DW_TAG_variable` out-of-line definition...

Sorry for the confusion. I wasn't referring to that code snippet.

Also, I must've not looked at the GCC dwarfdump output carefully enough. Compiling two CUs that include a class with a `static constexpr` member will produce two separate definitions and two entries in `.debug_pubnames`:
```
.debug_pubnames                                                                                            
global die-in-sect 0x00000070, cu-in-sect 0x0000000c, die-in-cu 0x00000070, cu-header-in-sect 0x00000000 'Foo::val2'                                                                                                   
global die-in-sect 0x0000007b, cu-in-sect 0x0000000c, die-in-cu 0x0000007b, cu-header-in-sect 0x00000000 'main'                                                                                                       
global die-in-sect 0x00000118, cu-in-sect 0x000000b4, die-in-cu 0x00000070, cu-header-in-sect 0x000000a8 'Foo::val2'                                                                                                   
global die-in-sect 0x00000128, cu-in-sect 0x000000b4, die-in-cu 0x00000080, cu-header-in-sect 0x000000a8 'Foo::func'               
```

So to answer Greg's question, yes we will potentially produce multiple variable definition DIEs. One for each CU that the header was included in

clayborg commented:
> > The DWARFASTParserClang.cpp will try to create the class from the DWARF for the class definition. You will need to find the DW_TAG_variable when we are creating the static field if there is no DW_AT_const_value in the DW_TAG_member. But we also need to support the DW_AT_const_value being in the DW_TAG_member since older DWARF will be emitted like this.
> 
> That's 100% correct. I was thinking, before [this block](https://github.com/llvm/llvm-project/blob/8b91de5d6a3f98dcc00bbd286e339e512f7e3682/lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp#L2909-L2919) where we check for the existence of a `const_value_form`, we could try to look for the definition and take the constant off of that.
> 
> What's interesting is that with this patch, the expression evaluator successfully finds the `DW_TAG_variable`s which have a location attribute but not if they have a constant instead of a location. It's probably some logic that assumes statics always have a location

The DWARFASTParserClang, with the current state of things, will automatically add the const value initializer to the clang AST field. See `DWARFASTParserClang::ParseSingleMember(...)` around the `// Handle static members` around lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:2882. The code extracts the `llvm::Expected<llvm::APInt> const_value_or_err` and then calls `TypeSystemClang::SetIntegerInitializerForVariable(v, *const_value_or_err);` to set the constant value of the static member. 

I think the expression parser knows how to grab this value if it is in a static member variable. If this isn't there, it assumes there is a global variable that backs it and that we will be able to find the location of this variable in memory. The expression parser will ask for the address of this value during expression evaluation when it resolves the symbols.

> 
> > Are we going to emit a DW_AT_const_expr now in the DW_TAG_member? If so, then we will know that we need to look for the DW_TAG_variable. I don't think clang emitted the DW_AT_const_expr attribute before.
> 
> That wasn't part of this patch. But would make sense to add (i've noticed GCC adds that attribute)

It would be nice to add this as a way to indicate this is a constexpr and that we need to do something special with it.

Is there anyway we can just leave the `DW_AT_const_value` in the `DW_TAG_member` and then have the `DW_TAG_variable` point to the `DW_TAG_member` using a `DW_AT_specification` or `DW_AT_abstract_origin`? My guess this isn't great DWARF where we have a `DW_TAG_variable` have a specification or abstract origin that points to a different DWARF tag.

Or maybe we can include the DW_AT_const_value in both places?

Michael137 commented:
> The DWARFASTParserClang, with the current state of things, will automatically add the const value initializer to the clang AST field. See `DWARFASTParserClang::ParseSingleMember(...)` around the `// Handle static members` around lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:2882. The code extracts the `llvm::Expected<llvm::APInt> const_value_or_err` and then calls `TypeSystemClang::SetIntegerInitializerForVariable(v, *const_value_or_err);` to set the constant value of the static member. 
> 
> 
> 
> I think the expression parser knows how to grab this value if it is in a static member variable. If this isn't there, it assumes there is a global variable that backs it and that we will be able to find the location of this variable in memory. The expression parser will ask for the address of this value during expression evaluation when it resolves the symbols.

Yup, what i was trying to explain is that with the current patch this will break because the expression evaluator neither finds the constant on the declaration (since we removed it) nor via symbol resolution (since they just won't exist for constants).

My point was just that if we are to drop the constant from the declaration, we'll have to amend that DWARFASTParserClang logic that you linked so it finds the correct variable definition and take the constant off of that.

> It would be nice to add this as a way to indicate this is a constexpr and that we need to do something special with it.

I agree, though i think that can be done in isolation as a follow-up PR.

> Is there anyway we can just leave the `DW_AT_const_value` in the `DW_TAG_member` and then have the `DW_TAG_variable` point to the `DW_TAG_member` using a `DW_AT_specification` or `DW_AT_abstract_origin`? My guess this isn't great DWARF where we have a `DW_TAG_variable` have a specification or abstract origin that points to a different DWARF tag.

The variable definition *will* have a specification back to the declaration.

> Or maybe we can include the DW_AT_const_value in both places?

The ask to drop the constant off of the declaration comes from the DWARFParallelLinker work where it was causing non-deterministic output. But @dwblaikie @avl-llvm will know more about that

avl-llvm commented:
> > Or maybe we can include the DW_AT_const_value in both places?
> 
> The ask to drop the constant off of the declaration comes from the DWARFParallelLinker work where it was causing non-deterministic output. But @dwblaikie @avl-llvm will know more about that

I do not have a strong opinion which variant is better : 1) remove DW_AT_const_value from DW_TAG_member and put DW_AT_const_value into newly created DW_TAG_variable(which does not have a location). or 2) always put  DW_AT_const_value in DW_TAG_member.

The problem which parallel DWARFLinker met is that DW_TAG_member from one compilation unit has DW_AT_const_value while the same DW_TAG_member from another compilation unit does not have DW_AT_const_value - https://github.com/llvm/llvm-project/pull/68721#issuecomment-1763409701.

If there is a solution when we could always put DW_AT_const_value into the DW_TAG_member (so that DW_TAG_member from different compile units have DW_AT_const_value) - then that would be also a good solution(and we would not need to create additional variable.)

Both solutions 1) and 2) are good for parallel DWARFLinker if they result in equal DWARF in different compile units.

dwblaikie commented:
> 2) always put DW_AT_const_value in DW_TAG_member.

My understanding is that this is not possible. Dependent initializer expressions can't be evaluated in all cases - we're only allowed to evaluate them in the places the language allows us to, otherwise we might produce the wrong answer.

For non-dependent initializer expressions I think we could produce the answer in all cases.

And if we want type descriptions that are consistent (valuable for use with Type Units, and with @avl-llvm's DWARFLinker work) and we want to include static member variables in those descriptions, then we could put the constant value in the declaration of the member in cases where the initializer is non-dependent. But we'd still have to put it out in a definition in the dependent cases (if we want consistency).

And if we have to put it out in a separate definition DIE anyway - probably good to do that consistently so there's fewer special cases?

Admittedly there are other reasons type definitions are inconsistent (eg: implicit special members, nested types, and member function template specializations (& I guess static member variable template specializations)) - and we could move static variables out of the authoritative type definitions the same way we do for those cases. We can see this in type units (the type unit never has these entities in it, but the skeleton unit that references the type unit does have these features) - I'd expect something like that to be what @avl-llvm will want to do with DWARFLinker - though without type units in input, I'm not sure how easily it'll be to determine that those are the variant parts that shouldn't go in the canonical type definition...

avl-llvm commented:
> > 2. always put DW_AT_const_value in DW_TAG_member.
> 
> My understanding is that this is not possible. Dependent initializer expressions can't be evaluated in all cases - we're only allowed to evaluate them in the places the language allows us to, otherwise we might produce the wrong answer.
> 
> For non-dependent initializer expressions I think we could produce the answer in all cases.
> 
> And if we want type descriptions that are consistent (valuable for use with Type Units, and with @avl-llvm's DWARFLinker work) and we want to include static member variables in those descriptions, then we could put the constant value in the declaration of the member in cases where the initializer is non-dependent. But we'd still have to put it out in a definition in the dependent cases (if we want consistency).
> 
> And if we have to put it out in a separate definition DIE anyway - probably good to do that consistently so there's fewer special cases?
> 
yep. Thanks. So it seems we have to drop DW_AT_const_value from DW_TAG_member if we would like to have consistent type descriptions.

> Admittedly there are other reasons type definitions are inconsistent (eg: implicit special members, nested types, and member function template specializations (& I guess static member variable template specializations)) - and we could move static variables out of the authoritative type definitions the same way we do for those cases. We can see this in type units (the type unit never has these entities in it, but the skeleton unit that references the type unit does have these features) - I'd expect something like that to be what @avl-llvm will want to do with DWARFLinker - though without type units in input, I'm not sure how easily it'll be to determine that those are the variant parts that shouldn't go in the canonical type definition...

Speaking for the other cases(It is probably out of the scoop of this patch) Are you saying that f.e. nested types or member function template specializations may look different in different compile units?

zmodem commented:
This causes Clang to assert:

```
      llvm/lib/IR/Metadata.cpp:689:
      void llvm::MDNode::resolve(): Assertion `isUniqued() && "Expected this to be uniqued"' failed.
```

See https://bugs.chromium.org/p/chromium/issues/detail?id=1500262#c1 for a reproducer.

I'll revert to green.

Michael137 commented:
> This causes Clang to assert:
> 
> ```
>       llvm/lib/IR/Metadata.cpp:689:
>       void llvm::MDNode::resolve(): Assertion `isUniqued() && "Expected this to be uniqued"' failed.
> ```
> 
> See https://bugs.chromium.org/p/chromium/issues/detail?id=1500262#c1 for a reproducer.
> 
> I'll revert to green.

Thanks for the repro. Reduced it to:
```
struct Foo;                                                 
template <bool c> using Int = int;                          
class Bar {                                                 
  template <typename T> static const bool Constant = false; 
  Int<Constant<Foo>> Val;                                   
  Bar();                                                    
};                                                          
Bar::Bar() {}                                               
```
when compiling with `-debug-info-kind=limited`.

The issue was that `EmitGlobalVariable` can cause creation of `DINodes`, which, if using limited debug-info, will be temporary. But the emission of the global variables is done *after* we've uniqued the temporaries. So the fix is really to just do the `EmitGlobalVariable` at the beginning of `CGDebugInfo::finalize`. Will submit a new PR with that change and a new test-case for this
